File: backend/src/providers/sharepoint.rs
use anyhow::{anyhow, Result};
use futures_util::StreamExt;
use tokio::fs::OpenOptions;
use tokio::io::AsyncWriteExt;

use crate::models::FileInfo;
use super::{apply_speed_limit, host_matches, unsupported_error, ProgressUpdate, Provider, ProviderDefaults};

pub struct SharePointProvider;

impl SharePointProvider {
    pub fn matches(url: &str) -> bool {
        host_matches(url, &["onedrive.live.com", "1drv.ms"])
            || reqwest::Url::parse(url)
                .ok()
                .and_then(|parsed| parsed.host_str().map(str::to_ascii_lowercase))
                .map(|host| host.ends_with("sharepoint.com"))
                .unwrap_or(false)
    }

    fn auth_required_error() -> anyhow::Error {
        unsupported_error("OneDrive/SharePoint")
    }

    fn direct_download_error() -> anyhow::Error {
        unsupported_error("OneDrive/SharePoint")
    }

    fn build_download_url(url: &str) -> Result<reqwest::Url> {
        let mut parsed = reqwest::Url::parse(url)
            .map_err(|_| anyhow!("URL do OneDrive/SharePoint inválida: {url}"))?;

        let mut pairs = parsed
            .query_pairs()
            .map(|(key, value)| (key.into_owned(), value.into_owned()))
            .collect::<Vec<_>>();

        let mut has_download = false;
        for (key, value) in &mut pairs {
            if key == "download" {
                *value = "1".to_string();
                has_download = true;
            }
        }
        if !has_download {
            pairs.push(("download".to_string(), "1".to_string()));
        }

        parsed.set_query(None);
        {
            let mut query = parsed.query_pairs_mut();
            for (key, value) in pairs {
                query.append_pair(&key, &value);
            }
        }

        Ok(parsed)
    }

    fn looks_like_login(url: &reqwest::Url) -> bool {
        match url.domain() {
            Some(domain) => {
                domain.eq_ignore_ascii_case("login.microsoftonline.com")
                    || domain.eq_ignore_ascii_case("login.live.com")
            }
            None => false,
        }
    }

    fn is_binary_response(resp: &reqwest::Response) -> bool {
        resp.headers().get("content-disposition").is_some()
            || resp
                .headers()
                .get("content-type")
                .and_then(|value| value.to_str().ok())
                .map(|value| {
                    let lower = value.to_ascii_lowercase();
                    !lower.contains("text/html") && !lower.contains("application/json")
                })
                .unwrap_or(false)
    }

    fn decode_path_segment(value: &str) -> String {
        let bytes = value.as_bytes();
        let mut decoded = Vec::with_capacity(bytes.len());
        let mut idx = 0;

        while idx < bytes.len() {
            if bytes[idx] == b'%' && idx + 2 < bytes.len() {
                let hex = &value[idx + 1..idx + 3];
                if let Ok(num) = u8::from_str_radix(hex, 16) {
                    decoded.push(num);
                    idx += 3;
                    continue;
                }
            }

            if bytes[idx] == b'+' {
                decoded.push(b' ');
            } else {
                decoded.push(bytes[idx]);
            }
            idx += 1;
        }

        String::from_utf8_lossy(&decoded).into_owned()
    }

    fn filename_from_response(resp: &reqwest::Response) -> Option<String> {
        let disposition = resp.headers().get("content-disposition")?.to_str().ok()?;
        for part in disposition.split(';') {
            let trimmed = part.trim();
            if let Some(value) = trimmed.strip_prefix("filename*=") {
                let value = value.split("''").last().unwrap_or(value);
                let decoded = Self::decode_path_segment(value.trim_matches('"'));
                if !decoded.is_empty() {
                    return Some(decoded);
                }
            }
            if let Some(value) = trimmed.strip_prefix("filename=") {
                let decoded = Self::decode_path_segment(value.trim_matches('"'));
                if !decoded.is_empty() {
                    return Some(decoded);
                }
            }
        }
        None
    }

    fn fallback_filename(url: &reqwest::Url) -> String {
        let candidate = url
            .path_segments()
            .and_then(|segments| segments.last())
            .filter(|segment| !segment.is_empty())
            .map(Self::decode_path_segment)
            .filter(|segment| !segment.is_empty())
            .unwrap_or_else(|| "arquivo_onedrive".to_string());

        <Self as ProviderDefaults>::safe_filename(&candidate, "arquivo_onedrive")
    }

    async fn request_public_download(url: &str) -> Result<reqwest::Response> {
        let client = <Self as ProviderDefaults>::http_client()?;
        let download_url = Self::build_download_url(url)?;
        Ok(client.get(download_url).send().await?.error_for_status()?)
    }

    async fn open_download_response(
        url: &str,
        dest_path: &str,
        existing_bytes: u64,
    ) -> Result<(reqwest::Response, bool)> {
        let client = <Self as ProviderDefaults>::http_client()?;
        let download_url = Self::build_download_url(url)?;

        if existing_bytes == 0 {
            return Ok((client.get(download_url).send().await?.error_for_status()?, false));
        }

        let ranged = client
            .get(download_url.clone())
            .header("Range", format!("bytes={existing_bytes}-"))
            .send()
            .await?;

        if ranged.status() == reqwest::StatusCode::PARTIAL_CONTENT {
            return Ok((ranged, true));
        }

        if matches!(
            ranged.status(),
            reqwest::StatusCode::BAD_REQUEST | reqwest::StatusCode::RANGE_NOT_SATISFIABLE
        ) {
            let _ = tokio::fs::remove_file(dest_path).await;
            let fresh = client.get(download_url).send().await?.error_for_status()?;
            return Ok((fresh, false));
        }

        Ok((ranged.error_for_status()?, false))
    }
}

impl ProviderDefaults for SharePointProvider {}

impl Provider for SharePointProvider {
    fn name(&self) -> &str { "OneDrive" }

    fn get_file_info<'a>(
        &'a self,
        url: &'a str,
    ) -> std::pin::Pin<Box<dyn std::future::Future<Output = Result<FileInfo>> + Send + 'a>> {
        Box::pin(async move {
            let resp = Self::request_public_download(url).await?;

            if Self::looks_like_login(resp.url()) {
                return Err(Self::auth_required_error());
            }

            if !Self::is_binary_response(&resp) {
                return Err(Self::direct_download_error());
            }

            let filename = Self::filename_from_response(&resp)
                .unwrap_or_else(|| Self::fallback_filename(resp.url()));
            let size = resp.content_length().unwrap_or(0);
            let mime_type = resp
                .headers()
                .get("content-type")
                .and_then(|value| value.to_str().ok())
                .map(|value| value.to_string());

            Ok(FileInfo {
                filename,
                size,
                mime_type,
                is_folder: false,
                children: None,
                ..Default::default()
            })
        })
    }

    fn download<'a>(
        &'a self,
        url: &'a str,
        dest_path: &'a str,
        speed_limit_bps: super::SpeedLimitBps,
        _parallel_parts: usize,
        _selected_children: Option<Vec<String>>,
        progress_tx: tokio::sync::mpsc::Sender<ProgressUpdate>,
    ) -> std::pin::Pin<Box<dyn std::future::Future<Output = Result<u64>> + Send + 'a>> {
        Box::pin(async move {
            let existing_bytes = tokio::fs::metadata(dest_path)
                .await
                .ok()
                .filter(|meta| meta.is_file())
                .map(|meta| meta.len())
                .unwrap_or(0);

            let (resp, resumed) = Self::open_download_response(url, dest_path, existing_bytes).await?;

            if Self::looks_like_login(resp.url()) {
                return Err(Self::auth_required_error());
            }

            if !Self::is_binary_response(&resp) {
                return Err(Self::direct_download_error());
            }

            let total = if resumed {
                <Self as ProviderDefaults>::response_total_bytes(&resp, existing_bytes)
            } else {
                resp.content_length().unwrap_or(0)
            };

            let mut file = if resumed {
                OpenOptions::new().create(true).append(true).open(dest_path).await?
            } else {
                tokio::fs::File::create(dest_path).await?
            };

            let mut downloaded = if resumed { existing_bytes } else { 0u64 };
            let mut session_downloaded = 0u64;
            let started_at = tokio::time::Instant::now();
            let mut stream = resp.bytes_stream();

            while let Some(chunk) = stream.next().await {
                let chunk = chunk?;
                file.write_all(&chunk).await?;
                let chunk_len = chunk.len() as u64;
                downloaded += chunk_len;
                session_downloaded += chunk_len;

                let _ = progress_tx.send(ProgressUpdate {
                    bytes_downloaded: downloaded,
                    total_bytes: total,
                    child_path: None,
                    child_filename: None,
                    child_bytes_downloaded: None,
                    child_total_bytes: None,
                    child_speed_bps: None,
                    child_eta_secs: None,
                }).await;

                apply_speed_limit(started_at, session_downloaded, &speed_limit_bps).await;
            }

            file.flush().await?;
            Ok(downloaded)
        })
    }
}

#[cfg(test)]
#[path = "tests/sharepoint.rs"]
mod tests;

File: backend-go/internal/providers/sharepoint.go
package providers

import (
	"fmt"
	"net/http"
	"net/url"
	"strings"
)

// Portado de backend/src/providers/sharepoint.rs — OneDrive/SharePoint public links.

type SharePointProvider struct{}

func (p SharePointProvider) Matches(rawURL string) bool {
	if HostMatches(rawURL, []string{"onedrive.live.com", "1drv.ms"}) {
		return true
	}
	u := ParseURL(rawURL)
	if u == nil {
		return false
	}
	host := strings.ToLower(u.Hostname())
	return strings.HasSuffix(host, "sharepoint.com")
}

func (p SharePointProvider) BuildDownloadURL(rawURL string) (*url.URL, error) {
	parsed, err := url.Parse(rawURL)
	if err != nil {
		return nil, fmt.Errorf("URL do OneDrive/SharePoint inválida: %s", rawURL)
	}
	q := parsed.Query()
	q.Set("download", "1")
	parsed.RawQuery = q.Encode()
	return parsed, nil
}

func (p SharePointProvider) LooksLikeLogin(u *url.URL) bool {
	host := strings.ToLower(u.Hostname())
	return host == "login.microsoftonline.com" || host == "login.live.com"
}

func (p SharePointProvider) IsBinaryResponse(resp *http.Response) bool {
	if resp.Header.Get("Content-Disposition") != "" {
		return true
	}
	ct := strings.ToLower(resp.Header.Get("Content-Type"))
	if ct == "" {
		return false
	}
	return !strings.Contains(ct, "text/html") && !strings.Contains(ct, "application/json")
}

func sharepointDecodePathSegment(value string) string {
	// decode %XX and + -> space, similar to Rust
	var decoded strings.Builder
	for i := 0; i < len(value); {
		if value[i] == '%' && i+2 < len(value) {
			hex := value[i+1 : i+3]
			var v int
			if _, err := fmt.Sscanf(hex, "%02x", &v); err == nil {
				decoded.WriteByte(byte(v))
				i += 3
				continue
			}
		}
		if value[i] == '+' {
			decoded.WriteByte(' ')
		} else {
			decoded.WriteByte(value[i])
		}
		i++
	}
	return decoded.String()
}

func (p SharePointProvider) FilenameFromResponse(resp *http.Response) *string {
	cd := resp.Header.Get("Content-Disposition")
	if cd == "" {
		return nil
	}
	for _, part := range strings.Split(cd, ";") {
		trimmed := strings.TrimSpace(part)
		if strings.HasPrefix(trimmed, "filename*=") {
			value := trimmed[len("filename*="):]
			if idx := strings.LastIndex(value, "''"); idx >= 0 {
				value = value[idx+2:]
			}
			value = strings.Trim(value, `"`)
			decoded := sharepointDecodePathSegment(value)
			if decoded != "" {
				return &decoded
			}
		}
		if strings.HasPrefix(trimmed, "filename=") {
			value := strings.Trim(trimmed[len("filename="):], `"`)
			decoded := sharepointDecodePathSegment(value)
			if decoded != "" {
				return &decoded
			}
		}
	}
	return nil
}

func (p SharePointProvider) FallbackFilename(u *url.URL) string {
	segs := strings.Split(strings.Trim(u.Path, "/"), "/")
	var candidate string
	if len(segs) > 0 {
		last := segs[len(segs)-1]
		if last != "" {
			candidate = sharepointDecodePathSegment(last)
		}
	}
	if candidate == "" {
		candidate = "arquivo_onedrive"
	}
	return SanitizeFilename(candidate, "arquivo_onedrive")
}

func (p SharePointProvider) RequestPublicDownload(client *http.Client, rawURL string) (*http.Response, error) {
	downloadURL, err := p.BuildDownloadURL(rawURL)
	if err != nil {
		return nil, err
	}
	req, _ := http.NewRequest("GET", downloadURL.String(), nil)
	resp, err := client.Do(req)
	if err != nil {
		return nil, err
	}
	if resp.StatusCode >= 400 {
		return nil, fmt.Errorf("OneDrive/SharePoint HTTP %d", resp.StatusCode)
	}
	return resp, nil
}

func (p SharePointProvider) FetchFileInfo(client *http.Client, rawURL string) (string, uint64, *string, error) {
	resp, err := p.RequestPublicDownload(client, rawURL)
	if err != nil {
		return "", 0, nil, err
	}
	defer resp.Body.Close()
	if p.LooksLikeLogin(resp.Request.URL) {
		return "", 0, nil, fmt.Errorf("UNSUPPORTED:OneDrive/SharePoint:Link não suportado pelo fluxo atual do OneDrive/SharePoint")
	}
	if !p.IsBinaryResponse(resp) {
		return "", 0, nil, fmt.Errorf("UNSUPPORTED:OneDrive/SharePoint:Link não suportado pelo fluxo atual do OneDrive/SharePoint")
	}
	var filename string
	if v := p.FilenameFromResponse(resp); v != nil {
		filename = *v
	} else {
		filename = p.FallbackFilename(resp.Request.URL)
	}
	var size uint64
	if cl := resp.Header.Get("Content-Length"); cl != "" {
		fmt.Sscanf(cl, "%d", &size)
	}
	var mime *string
	if ct := resp.Header.Get("Content-Type"); ct != "" {
		mime = &ct
	}
	return SanitizeFilename(filename, "arquivo_onedrive"), size, mime, nil
}
